src/libtrx/game/objects/general/trapdoor.c

```diff
@@ -103,14 +117,92 @@ static void M_Control(const int16_t item_num)
     if (Item_IsTriggerActive(item)) {
         if (item->current_anim_state == TRAPDOOR_STATE_CLOSED) {
             item->goal_anim_state = TRAPDOOR_STATE_OPEN;
+            M_DropStack(item);
         }
     } else {
         if (item->current_anim_state == TRAPDOOR_STATE_OPEN) {
             item->goal_anim_state = TRAPDOOR_STATE_CLOSED;
         }
     }
+
     Item_Animate(item);
 }
 
+void M_DropStack(const ITEM *const item)
+{
+    VECTOR *pos = Vector_Create(sizeof(XYZ_32));
+    Trapdoor_GetSectorPositions(item, pos);
+    for (int32_t i = 0; i < pos->count; i++) {
+        MovableBlock_DropStack(
+            item->pos.y, *(const XYZ_32 *)Vector_Get(pos, i), item->room_num);
+    }
+    Vector_Free(pos);
+    pos = nullptr;
+}
+
+static int32_t floor_div(int32_t v, int32_t d)
+{
+    return (v >= 0) ? v / d : -((-v + d - 1) / d);
+}
```

Comment:
This actually appears more than once so I'd suggest moving it into `math/` as a public function, `Math_FloorDiv`. Args can be const BTW